README.md
# ASProgressHud
![Swift 4.0](https://img.shields.io/badge/Swift-4.0-orange.svg?maxAge=3600)
[![Platform](https://img.shields.io/cocoapods/p/ASProgressHud.svg?style=flat)](http://cocoapods.org/pods/ASProgressHud)
[![Version](https://img.shields.io/cocoapods/v/ASProgressHud.svg?style=flat)](http://cocoapods.org/pods/ASProgressHud)
[![License](https://img.shields.io/cocoapods/l/ASProgressHud.svg?style=flat)](http://cocoapods.org/pods/ASProgressHud)
[![Apps](https://img.shields.io/cocoapods/at/ASProgressHud.svg?style=flat)](http://cocoapods.org/pods/ASProgressHud)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![BuddyBuild](https://dashboard.buddybuild.com/api/statusImage?appID=58c125ce8fe03c010009b1a5&branch=master&build=latest)](https://dashboard.buddybuild.com/apps/56f4f461ae27cb01000b366d/build/latest?branch=master)
[![codebeat badge](https://codebeat.co/badges/ac582740-1971-481c-a11d-3c62f6f8e618)](https://codebeat.co/projects/github-com-andr3a88-asprogresshud-master)

## Requirements

* Swift 5
* XCode 10.2
* iOS >= 9

## Installation

ASProgressHud is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "ASProgressHud"
```

## Changelog
See the [changelog](CHANGELOG.md) file. 


## Description
ASProgressHud displays a customized loading view. The loader is created with an UIImageView animated with png images.

![Screenshot](preview.gif)

## Usage

To run the example project, clone the repo, and run `pod install` from the Example directory first.

#### Default HUD

```swift
// Show
ASProgressHud.showHUDAddedTo(self.view, animated: true, type: .default)

// Hide
ASProgressHud.hideHUDForView(self.view, animated: true)
```

#### Custom HUD

```swift
// Import loader images in png format, with name like loader\_custom\_00.png, loader\_custom\_01.png, etc...

// Create HudProperty
HudProperty(prefixName: "loader_custom", frameNumber: 18)

// Show
ASProgressHud.showCustomHUDAddedTo(self.view, animated: true, hudProperty: hudProperty)

// Hide
ASProgressHud.hideHUDForView(self.view, animated: true)
```

## Resources

* [Preloaders.net](http://preloaders.net/)
* [Gif and aPng splitter](http://animizer.net/en/gif-apng-splitter)

## Author

Andrea, Twitter @andrea_steva

## License

ASProgressHud is available under the MIT license. See the LICENSE file for more info.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

### [1.0.7](https://github.com/andr3a88/ASProgressHud/tree/1.0.7)

Support to Swift 5

### [1.0.6](https://github.com/andr3a88/ASProgressHud/tree/1.0.6)

Support to Swift 4.2

### [1.0.5](https://github.com/andr3a88/ASProgressHud/tree/1.0.5)

Support to Swift 4.2

### [1.0.4](https://github.com/andr3a88/ASProgressHud/tree/1.0.4)

Support to Swift 4

### [1.0.3](https://github.com/andr3a88/ASProgressHud/tree/1.0.3)

Fix issue on animation duration.

### [1.0.2](https://github.com/andr3a88/ASProgressHud/tree/1.0.2)

Updated README and minor issue.

### [1.0.1](https://github.com/andr3a88/ASProgressHud/tree/1.0.1)

Updated README and minor issue.

### [1.0.0](https://github.com/andr3a88/ASProgressHud/tree/1.0.0)

Converted to Swift 3.0.

### [0.0.3](https://github.com/andr3a88/ASProgressHud/tree/0.0.3)

Converted to Swift 2.2.

### [0.0.2](https://github.com/andr3a88/ASProgressHud/tree/0.0.2)

Release with Cocoapods 1.0 Beta

### [0.0.1](https://github.com/andr3a88/ASProgressHud/tree/0.0.1)

Swift 2.0 version.


Example/Tests/Tests.swift
import UIKit
import XCTest
@testable import ASProgressHud

class Tests: XCTestCase {
    
    func testHudPropertyInitializer() {
        
        var property = HudProperty(prefixName: "_prefix", frameNumber: 30)
        
        XCTAssertEqual(property.frameNumber, 30)
        XCTAssertEqual(property.prefixName, "_prefix")
        
        property = HudProperty(prefixName: "image_", frameNumber: 40, size: 10.0, animationDuration: 1.0, backgroundColor: UIColor.red)
        
        XCTAssertEqual(property.frameNumber, 40)
        XCTAssertEqual(property.prefixName, "image_")
        XCTAssertEqual(property.size, 10.0)
        XCTAssertEqual(property.backgroundColor, UIColor.red)
    }
    
    func testHudType() {
        
        var type = HudType.default
        XCTAssertEqual(type.properties.prefixName, "default")
        XCTAssertEqual(type.properties.frameNumber, 18)
        
        type = HudType.flag
        XCTAssertEqual(type.properties.prefixName, "flag")
        XCTAssertEqual(type.properties.frameNumber, 20)
        
        type = HudType.google
        XCTAssertEqual(type.properties.prefixName, "google")
        XCTAssertEqual(type.properties.frameNumber, 30)
    }
    
    func testProgressHudInitializer() {
        
        let frame = CGRect(x: 0, y: 0, width: 100, height: 100)
        let property = HudProperty(prefixName: "google", frameNumber: 30)
        
        var hud = ASProgressHud(frame: frame, type: HudType.default)
        XCTAssertEqual(hud.frame, frame)
        
        hud = ASProgressHud(frame: frame, hudProperty: property)
        XCTAssertEqual(hud.frame, frame)
    }
}


Example/ASProgressHud/AppDelegate.swift
//
//  AppDelegate.swift
//  ASProgressHud
//
//  Created by Andrea on 02/17/2016.
//  Copyright (c) 2016 Andrea. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
    
    var window: UIWindow?
    
    private func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }
    
    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }
    
    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }
    
    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }
    
    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }
    
    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }
    
}


Example/ASProgressHud/ExampleViewController.swift
//
//  ExampleViewController.swift
//  ASProgressHud
//
//  Created by Andrea on 02/17/2016.
//  Copyright (c) 2016 Andrea. All rights reserved.
//

import UIKit
import ASProgressHud

class ExampleViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    
    @IBAction func showWhiteHUD() {
        
        ASProgressHud.showHUDAddedTo(self.view, animated: true, type: .default)
        print("Hud SHOW")
        
        self.hideAfterDelay()
    }
    
    @IBAction func showGoogleHUD() {
        
        ASProgressHud.showHUDAddedTo(self.view, animated: true, type: .google)
        print("Hud SHOW")
        
        self.hideAfterDelay()
    }
    
    @IBAction func showFlagHUD() {
        
        ASProgressHud.showHUDAddedTo(self.view, animated: true, type: .flag)
        print("Hud SHOW")
        
        self.hideAfterDelay()
    }
    
    @IBAction func showCustomHUD() {
        
        let property = HudProperty(prefixName: "coffee", frameNumber: 13, size: 70, animationDuration: 0.2, backgroundColor: UIColor.yellow)
        ASProgressHud.showCustomHUDAddedTo(self.view, animated: true, property: property)
        print("Hud SHOW")
        
        self.hideAfterDelay()
    }
    
    func hideAfterDelay() {
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(3)) {
            ASProgressHud.hideHUDForView(self.view, animated: true)
            print("Hud HIDE")
        }
    }
    
}



Pod/Classes/ASProgressHud.swift
//
//  ASProgress.swift
//  Pods
//
//  Created by Andrea Stevanato on 17/02/16.
//
//

/**
*  Specify the property for the Hud.
*/
public struct HudProperty {
    
    /// The images predix name (eg. For a image with format "my_loader_XX.png" with NN from 0 to 20 the predix name is "my:loader")
    public let prefixName: String
    /// The number of images used to animate the loader
    let frameNumber: Int
    /// The size for the loader. Defaut value 60 px
    let size: CGFloat
    /// The animation duration for a single cycles. Default value 1.0 second
    let animationDuration: TimeInterval
    /// An optional background color. The defuault value is clear color
    let backgroundColor: UIColor?
    
    fileprivate var mainBundle: Bool = false
    
    /**
     Specify the property for the Hud
     
     - parameter prefixName:        The images predix name (eg. For a image with format "my_loader_XX.png" with NN from 0 to 20 the predix name is "my:loader")
     - parameter frameNumber:       The number of images used to animate the loader
     - parameter size:              The size for the loader. Defaut value 60 px
     - parameter animationDuration: The animation duration for a single cycles. Default value 1.0 second
     - parameter backgroundColor:   An optional background color. The defuault value is clear color
     
     - returns: The property struct
     */
    public init(prefixName: String, frameNumber: Int, size: CGFloat = 60, animationDuration: TimeInterval = 1.0, backgroundColor: UIColor? = UIColor.clear) {
        self.prefixName = prefixName
        self.frameNumber = frameNumber
        self.size = size
        self.animationDuration = animationDuration
        self.backgroundColor = backgroundColor
    }
}

/// A set of custom Hud
public enum HudType: Int {
    case `default`, flag, google
    
    var properties: HudProperty {
        switch self {
        case .default:
            return HudProperty(prefixName: "default", frameNumber: 18)
        case .flag:
            return HudProperty(prefixName: "flag", frameNumber: 20)
        case .google:
            return HudProperty(prefixName: "google", frameNumber: 30)
        }
    }
}

/**
 *  ASProgressHud is UIView subclass
 */
open class ASProgressHud: UIView {
    
    fileprivate var useAnimation = true
    fileprivate var showStarted: Date?
    fileprivate var hudImageView: UIImageView?
    var removeFromSuperViewOnHide: Bool? = false
    
    // MARK: - Init
    
    public required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    /**
     Init with Progress Hud Type
     
     :param: frame frame
     :param: type  Progress Hud Type
     
     :returns: ASProgressHud istance
     */
    init(frame: CGRect, type: HudType) {
        super.init(frame: frame)
        
        let hudProperty = type.properties
        self.initializeWithHudProperty(hudProperty)
        
        self.customTypeBasedConfiguration(type)
        self.addSubview(self.hudImageView!)
    }
    
    /**
     Init with custom loader
     
     :param: frame       Frame
     :param: hudProperty Hud property
     
     :returns: ASProgressHud istance
     */
    init(frame: CGRect, hudProperty: HudProperty) {
        super.init(frame: frame)
        
        self.initializeWithHudProperty(hudProperty)
        self.addSubview(self.hudImageView!)
    }
    
    fileprivate func initializeWithHudProperty(_ hudProperty: HudProperty) {
        self.removeFromSuperViewOnHide = true
        self.backgroundColor = UIColor.black.withAlphaComponent(0.2)
        self.isOpaque = false
        self.alpha = 0.0
        self.contentMode = UIView.ContentMode.center
        self.autoresizingMask = [.flexibleTopMargin, .flexibleBottomMargin, .flexibleLeftMargin, .flexibleRightMargin]
        
        // Configure HudImageView
        self.hudImageView = UIImageView(frame: CGRect(x: 0, y: 0, width: hudProperty.size, height: hudProperty.size))
        self.hudImageView?.center = CGPoint(x: frame.size.width / 2, y: frame.size.height / 2)
        self.hudImageView?.contentMode = UIView.ContentMode.scaleAspectFit
        self.hudImageView?.backgroundColor = hudProperty.backgroundColor
        
        // Load images
        let imagesArray = self.loadImages(hudProperty)
        
        // Animation configuration
        self.hudImageView?.animationImages = imagesArray
        self.hudImageView?.animationDuration = hudProperty.animationDuration
        self.hudImageView?.animationRepeatCount = 0
        self.hudImageView?.clipsToBounds = true
        self.hudImageView?.layer.cornerRadius = 10.0
    }
    
    fileprivate func customTypeBasedConfiguration(_ type: HudType) {
        switch type {
        case .default:
            
            // Add a black shadow with radius offset and opacity
            self.hudImageView?.layer.shadowColor = UIColor.black.cgColor
            self.hudImageView?.layer.shadowOpacity = 0.4
            self.hudImageView?.layer.shadowRadius = 3.0
            self.hudImageView?.layer.shadowOffset = CGSize(width: 0, height: 3)
            self.hudImageView?.clipsToBounds = false
            
        default:
            break
        }
    }
    
    // MARK: - Utils
    
    fileprivate func loadImages(_ property: HudProperty) -> [UIImage] {

        // Load images
        var imageArray: [UIImage] = []
        
        for c in 0 ..< property.frameNumber {
            guard let image = UIImage(named: String(format: "%@_%02d.png", property.prefixName, c), in: property.mainBundle ? nil : self.podBundle(), compatibleWith: nil) else {
                assertionFailure("Cannot load the image")
                return []
            }
            imageArray.append(image)
        }
        
        return imageArray
    }
    
    fileprivate func podBundle() -> Bundle {
        
        let podBundle = Bundle(for: self.classForCoder)
        if let bundleURL = podBundle.url(forResource: "Resources", withExtension: "bundle") {
            
            if let bundle = Bundle(url: bundleURL) {
                return bundle
            } else {
                assertionFailure("Could not load the bundle")
            }
        } else {
            assertionFailure("Could not create a path to the bundle")
        }
        return podBundle
    }
    
    // MARK: - Class methods
    
    /**
    Creates a new HUD, adds it to provided view and shows it. The counterpart to this method is hideHUDForView:animated:.
    
    :note: This method sets `removeFromSuperViewOnHide`. The HUD will automatically be removed from the view hierarchy when hidden.
    
    :param: view     The view that the HUD will be added to
    :param: animated If set to YES the HUD will appear using the current animationType. If set to NO the HUD will not use animations while appearing.
    :param: type     The type for the HUD
    
    :returns: A reference to the created HUD.
    
    :see: hideHUDForView:animated:
    :see: animationType
    */
    @discardableResult
    public static func showHUDAddedTo(_ view: UIView, animated: Bool, type: HudType) -> ASProgressHud {
        let hud = ASProgressHud(frame: view.bounds, type: type)
        view.addSubview(hud)
        hud.show(animated)
        return hud
    }
    
    /**
     Creates a new HUD, adds it to provided view and shows it. The counterpart to this method is hideHUDForView:animated:.
     
     :note: This method sets `removeFromSuperViewOnHide`. The HUD will automatically be removed from the view hierarchy when hidden.
     
     :param: view     The view that the HUD will be added to
     :param: animated If set to YES the HUD will appear using the current animationType. If set to NO the HUD will not use animations while appearing.
     :param: type     The type for the HUD
     
     :returns: A reference to the created HUD.
     
     :see: hideHUDForView:animated:
     :see: animationType
     */
    @discardableResult
    public static func showCustomHUDAddedTo(_ view: UIView, animated: Bool, property: HudProperty) -> ASProgressHud {
        var hudProperty = property
        hudProperty.mainBundle = true
        let hud = ASProgressHud(frame: view.bounds, hudProperty: hudProperty)
        view.addSubview(hud)
        hud.show(animated)
        return hud
    }
    
    /**
     Finds the top-most HUD subview and hides it. The counterpart to this method is showHUDAddedTo:animated:.
     
     :note: This method sets `removeFromSuperViewOnHide`. The HUD will automatically be removed from the view hierarchy when hidden.
     
     :param: view     The view that is going to be searched for a HUD subview.
     :param: animated If set to YES the HUD will disappear using the current animationType. If set to NO the HUD will not use animations while disappearing.
     
     :returns: YES if a HUD was found and removed, NO otherwise.
     
     :see: showHUDAddedTo:animated:
     :see: animationType
     */
    @discardableResult
    public static func hideHUDForView(_ view: UIView, animated: Bool) -> Bool {
        let hud = self.HUDForView(view)
        
        if hud != nil {
            hud?.removeFromSuperViewOnHide = true
            hud?.hide(animated)
            return true
        }
        return false
    }
    
    /**
     Finds all the HUD subviews and hides them.
     
     :param: view     The view that is going to be searched for HUD subviews.
     :param: animated If set to YES the HUDs will disappear using the current animationType. If set to NO the HUDs will not use animations while disappearing.
     
     :returns: The number of HUDs found and removed.
     
     :see: hideHUDForView:animated:
     :see: animationType
     */
    public static func hideAllHUDsForView(_ view: UIView, animated: Bool) -> NSInteger {
        
        let huds = ASProgressHud.allHUDsForView(view)
        for hud in huds {
            hud.removeFromSuperViewOnHide = true
            hud.hide(animated)
        }
        return huds.count
    }
    
    fileprivate static func HUDForView(_ view: UIView) -> ASProgressHud? {
        
        let subviewsEnum = view.subviews as NSArray
        let array = subviewsEnum.reverseObjectEnumerator()
        
        for subView in array {
            if (subView as AnyObject).isKind(of: self) {
                return subView as? ASProgressHud
            }
        }
        return nil
    }
    
    fileprivate static func allHUDsForView(_ view: UIView) -> [ASProgressHud] {
        
        var huds: [ASProgressHud]? = []
        let subviews = view.subviews as Array
        for aView in subviews {
            if aView.isKind(of: self) {
                huds?.append(aView as! ASProgressHud)
            }
        }
        return huds!
    }
    
    // MARK: - Show & hide
    
    fileprivate func showUsingAnimation(_ animated: Bool) {
        // Cancel any scheduled hideDelayed: calls
        NSObject.cancelPreviousPerformRequests(withTarget: self)
        self.setNeedsDisplay()
        
        self.hudImageView?.startAnimating()
        
        self.showStarted = Date()
        
        // Fade in
        if animated == true {
            UIView.beginAnimations(nil, context: nil)
            UIView.setAnimationDuration(0.30)
            self.alpha = 1.0
            UIView.commitAnimations()
        } else {
            self.alpha = 1.0
        }
    }
    
    fileprivate func hideUsingAnimation(_ animated: Bool) {
        // Fade out
        if animated == true && showStarted != nil {
            UIView.beginAnimations(nil, context: nil)
            UIView.setAnimationDuration(0.30)
            UIView.setAnimationDelegate(self)
            UIView.setAnimationDidStop(#selector(ASProgressHud.animationFinished))
            // 0.02 prevents the hud from passing through touches during the animation the hud will get completely hidden
            // in the done method
            self.alpha = 0.02
            UIView.commitAnimations()
        } else {
            self.alpha = 0.0
            self.done()
        }
        self.showStarted = nil
    }
    
    fileprivate func show(_ animated: Bool) {
        assert(Thread.isMainThread, "ASProgressHud needs to be accessed on the main thread.")
        self.useAnimation = animated
        self.showUsingAnimation(useAnimation)
    }
    
    fileprivate func hide(_ animated: Bool) {
        assert(Thread.isMainThread, "ASProgressHud needs to be accessed on the main thread.")
        self.useAnimation = animated
        self.hideUsingAnimation(useAnimation)
    }
    
    // MARK: - Internal show & hide operations
    
    @objc func animationFinished() {
        self.done()
    }
    
    fileprivate func done() {
        NSObject.cancelPreviousPerformRequests(withTarget: self)
        self.alpha = 0.0
        self.hudImageView?.stopAnimating()
        
        if removeFromSuperViewOnHide == true {
            self.removeFromSuperview()
        }
    }
    
}
